Userland/Libraries/LibWeb/CSS/Parser/DeprecatedCSSParser.cpp

```diff
@@ -229,6 +230,125 @@ static StringView parse_custom_property_name(const StringView& value)
 }
 
+static StringView isolate_calc_expression(const StringView& value)
+{
+    if (!value.starts_with("calc(") || !value.ends_with(")"))
+        return {};
+    auto substring_length = value.length() - 5 - 1;
+    return value.substring_view(5, substring_length);
+}
+
+static CSS::CalculatedStyleValue::CalcSum parse_calc_expression(const StringView& expression_string)
+{
+    // First, tokenize
+
+    struct CalcToken {
+        enum class Type {
+            Undefined,
+            Number,
+            Unit,
+            Whitespace,
+            Plus,
+            Minus,
+            Asterisk,
+            Slash,
+            OpenBracket,
+            CloseBracket,
+        } type { Type::Undefined };
+        String value {};
+    };
+
+    size_t index = 0;
+    Vector<CalcToken> tokens;
+
+    char ch;
+    auto consume = [&index, &ch, &expression_string]() {
+        ch = expression_string[index];
+        index++;
+        return ch;
+    };
+
+    auto peek = [&index, &expression_string](int offset) {
+        auto new_index = index + offset - 1;
+        return expression_string[new_index];
+    };
+
+    while (index < expression_string.length()) {
+        consume();
+        // Number
+        if (((ch == '+' || ch == '-') && !isspace(peek(1))) || ch == '.' || isdigit(ch)) {
+            StringBuilder number;
+            while (ch == '+' || ch == '-' || ch == '.' || isdigit(ch)) {
+                number.append(ch);
+                consume();
+            }
+            index--;
+            tokens.append(CalcToken { CalcToken::Type::Number, number.to_string() });
+            continue;
+        }
+
+        // Single-character-tokens
+
+        if (isspace(ch)) {
+            tokens.append(CalcToken { CalcToken::Type::Whitespace });
+            continue;
+        }
+
+        if (ch == '%') {
+            tokens.append(CalcToken { CalcToken::Type::Unit, "%" });
+            continue;
+        }
+
+        if (ch == '+') {
+            tokens.append(CalcToken { CalcToken::Type::Plus });
+            continue;
+        }
+
+        if (ch == '-') {
+            tokens.append(CalcToken { CalcToken::Type::Minus });
+            continue;
+        }
+
+        if (ch == '*') {
+            tokens.append(CalcToken { CalcToken::Type::Asterisk });
+            continue;
+        }
+
+        if (ch == '/') {
+            tokens.append(CalcToken { CalcToken::Type::Slash });
+            continue;
+        }
+
+        if (ch == '(') {
+            tokens.append(CalcToken { CalcToken::Type::OpenBracket });
+            continue;
+        }
+
+        if (ch == ')') {
+            tokens.append(CalcToken { CalcToken::Type::CloseBracket });
+            continue;
+        }
+
+        // Unit
+        if (isalpha(ch)) {
+            StringBuilder string;
+            while (isalpha(ch)) {
+                string.append(ch);
+                consume();
+            }
+            index--;
```

Comment:
This copy is sad, see `GenericLexer::consume_while(isalpha)` which gives you a `StringView` instead.
less code _and_ less copy